src/attached_probe.h

```diff
@@ -40,84 +40,208 @@ class AttachedProbe {
+                    std::function<void()> cleanup);
+  int perf_event_fd_;
+  std::function<void()> usdt_sem_cleanup_;
+};
+
+class AttachedTracepointProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedTracepointProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog);
+  ~AttachedTracepointProbe() override;
+
+private:
+  AttachedTracepointProbe(int progfd,
+                          int perf_event_fd,
+                          std::string event_name,
+                          std::string probe_path);
+  int perf_event_fd_;
+  const std::string event_name_;
+  const std::string probe_path_;
+};
+
+class AttachedProfileProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedProfileProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog,
+      std::optional<int> pid);
+  ~AttachedProfileProbe() override;
+
+private:
+  AttachedProfileProbe(int progfd, std::vector<int> perf_event_fds);
+  std::vector<int> perf_event_fds_;
+};
+
+class AttachedIntervalProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedIntervalProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog,
+      std::optional<int> pid);
+  ~AttachedIntervalProbe() override;
+
+private:
+  AttachedIntervalProbe(int progfd, int perf_event_fd);
+  int perf_event_fd_;
+};
+
+class AttachedSoftwareProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedSoftwareProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog,
+      std::optional<int> pid);
+  ~AttachedSoftwareProbe() override;
+
+private:
+  AttachedSoftwareProbe(int progfd, std::vector<int> perf_event_fds);
+  std::vector<int> perf_event_fds_;
+};
+
+class AttachedHardwareProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedHardwareProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog,
+      std::optional<int> pid);
+  ~AttachedHardwareProbe() override;
+
+private:
+  AttachedHardwareProbe(int progfd, std::vector<int> perf_event_fds);
+  std::vector<int> perf_event_fds_;
+};
+
+class AttachedFentryProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedFentryProbe>> make(
+      const BpfProgram &prog);
+  ~AttachedFentryProbe() override;
+
+private:
+  AttachedFentryProbe(int progfd, int tracing_fd);
+  int tracing_fd_;
+};
+
+class AttachedRawtracepointProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedRawtracepointProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog);
+  ~AttachedRawtracepointProbe() override;
+
+private:
+  AttachedRawtracepointProbe(int progfd, int tracing_fd);
+  int tracing_fd_;
+};
+
+class AttachedIterProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedIterProbe>> make(
+      const BpfProgram &prog,
+      std::optional<int> pid);
+  ~AttachedIterProbe() override;
+
+  const int iter_link_fd;
```

Comment:
Similar to AttachedWatchpoingProbe, this could be moved to something like `get_link_fd()` on the base class. This gets quite clean for my follow-on work (as well as @Phoenix500526's), as we can slowly implement `get_link_fd()` on subclasses that are migrated to libbpf attachment.